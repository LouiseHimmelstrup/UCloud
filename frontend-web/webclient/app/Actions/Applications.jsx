import { Cloud } from "../../authentication/SDUCloudObject";
import { RECEIVE_APPLICATIONS, SET_LOADING, UPDATE_APPLICATIONS_PER_PAGE } from "../Reducers/Applications";

const receiveApplications = (applications) => ({
    type: RECEIVE_APPLICATIONS,
    applications
});

export const fetchApplications = () =>
    Cloud.get("/hpc/apps").then(({ response }) => {
        response.sort((a, b) =>
            a.prettyName.localeCompare(b.prettyName)
        );
        return receiveApplications(response);
    });

export const setLoading = (loading) => ({
    type: SET_LOADING,
    loading
});

export const toPage = (pageNumber) => ({
    type: TO_PAGE,
    pageNumber
});

export const updateApplications = (applications) => ({
    type: UPDATE_APPLICATIONS,
    applications
});

export const updateApplicationsPerPage = (applicationsPerPage) => ({
    type: UPDATE_APPLICATIONS_PER_PAGE,
    applicationsPerPage
});