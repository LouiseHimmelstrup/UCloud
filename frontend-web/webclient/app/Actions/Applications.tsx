import { Cloud } from "../../authentication/SDUCloudObject";
import { failureNotification } from "../UtilityFunctions";
import {
    RECEIVE_APPLICATIONS,
    SET_APPLICATIONS_LOADING,
    UPDATE_APPLICATIONS
} from "../Reducers/Applications";

import {
    Application,
    Action,
    SetLoadingAction,
    Page
} from "../Types";
import { emptyPage } from "../DefaultObjects";

interface ReceiveApplicationsAction extends Action { page: Page<Application> }
const receiveApplications = (page: Page<Application>): ReceiveApplicationsAction => ({
    type: RECEIVE_APPLICATIONS,
    page
});

export const fetchApplications = (page: number, itemsPerPage: number): Promise<ReceiveApplicationsAction> =>
    Cloud.get(`/hpc/apps?page=${page}&itemsPerPage=${itemsPerPage}`).then(({ response }: { response: Page<Application> }) => {
        return receiveApplications(response);
    }).catch(() => {
        failureNotification("An error occurred while retrieving applications.")
        return receiveApplications(emptyPage);
    });

/**
 * Sets the loading state for the applications component
 * @param {boolean} loading loading state for the applications component
 */
export const setLoading = (loading: boolean): SetLoadingAction => ({
    type: SET_APPLICATIONS_LOADING,
    loading
});

export const updateApplications = (page: Page<Application>): ReceiveApplicationsAction => ({
    type: UPDATE_APPLICATIONS,
    page
});