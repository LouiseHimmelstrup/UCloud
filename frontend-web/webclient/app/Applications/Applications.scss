.stream {
    height: 500px;
    overflow: auto;
}

.input-group > .form-control.time-selection {
    width: 100px;
}

.text-area-no-resize {
    resize: none;
}

.job-result-box {
    padding-bottom: 16px;
}

.cards.card-margin {
   margin-top: 20px
}