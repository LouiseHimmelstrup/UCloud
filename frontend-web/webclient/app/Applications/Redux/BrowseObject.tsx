import { LoadableContent, emptyLoadableContent } from "LoadableContent";
import { Page } from "Types";
import { ApplicationMetadata, WithAppMetadata } from "Applications";

export interface Type {
    applications: Map<string, Page<WithAppMetadata>>
    loading: boolean
}

export interface Wrapper {
    applicationsBrowse: Type
}

export const init = (): Wrapper => ({
    applicationsBrowse: {
        applications: new Map(),
        loading: false
    }
});