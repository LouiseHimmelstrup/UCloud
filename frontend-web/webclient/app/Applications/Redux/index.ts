import * as ViewObject from "./ViewObject";
import * as ViewReducer from "./ViewReducer";
import * as ApplicationObject from "./ApplicationsObject";
import * as ApplicationReducer from "./ApplicationsReducer";

export type Reducers = ViewReducer.Reducer & ApplicationReducer.Reducer;
export type Objects = ViewObject.Wrapper & ApplicationObject.Wrapper;

export function init(): Objects {
    return { 
        ...ViewObject.init(),
        ...ApplicationObject.init()
    }
}

export const reducers: Reducers = {
    applicationView: ViewReducer.default,
    applicationsBrowse: ApplicationReducer.default
};