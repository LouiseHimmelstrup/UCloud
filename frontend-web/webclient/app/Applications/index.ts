import { File } from "Files";
import { Page } from "Types";
import { match } from "react-router";
import PromiseKeeper from "PromiseKeeper";
import { History } from "history";
import { DetailedResultReduxObject, ApplicationReduxObject } from "DefaultObjects";

export type ApplicationsProps = ApplicationReduxObject & ApplicationsOperations;

export interface Analysis {
    status: string
    state: string
    jobId: string
    appName: string
    appVersion: string
    createdAt: number
    modifiedAt: number
    owner: string
}

export interface ApplicationsOperations {
    prioritizeApplicationSearch: () => void
    onErrorDismiss: () => void
    updatePageTitle: () => void
    setLoading: (loading: boolean) => void
    setFavoritesLoading: (loading: boolean) => void
    fetchApplications: (a: number, b: number) => void
    fetchFavorites: (a: number, b: number) => void
    receiveApplications: (applications: Page<Application>) => void
}

export interface AnalysesProps extends AnalysesStateProps, AnalysesOperations { }

export interface AnalysesStateProps {
    page: Page<Analysis>
    loading: boolean
    error?: string
}

export interface AnalysesOperations {
    onErrorDismiss: () => void
    updatePageTitle: (title: string) => void
    setLoading: (loading: boolean) => void
    fetchAnalyses: (itemsPerPage: number, pageNumber: number) => void
}

export interface AnalysesState {
    reloadIntervalId: number
}

export interface DetailedResultOperations {
    receivePage: (page: Page<File>) => void,
    setPageTitle: (jobId: string) => void
    setLoading: (loading: boolean) => void
    detailedResultError: (error: string) => void
    fetchPage: (jobId: string, pageNumber: number, itemsPerPage: number) => void
}

export interface DetailedResultProps extends DetailedResultReduxObject, DetailedResultOperations {
    match: match<{ jobId: string }>
    history: History
}

export interface Application {
    favorite: boolean
    owner: string
    createdAt: number
    modifiedAt: number
    description: ApplicationDescription
    tool: ApplicationTool
}

interface ApplicationTool {
    owner: string
    createdAt: number
    modifiedAt: number
    description: {
        info: ApplicationInfo
        container: string
        defaultNumberOfNodes: number
        defaultTasksPerNode: number
        defaultMaxTime: MaxTime
        requiredModules: string[]
        authors: string[]
        title: string
        description: string
        backend: string
    }
}

interface ApplicationInfo {
    name: string
    version: string
}

export interface ApplicationDescription {
    info: ApplicationInfo
    tool: ApplicationInfo
    authors: string[]
    title: string
    description: string
    invocation: any[]
    parameters: ApplicationParameter[]
    outputFileGlobs: string[]
}

export interface DetailedResultState {
    complete: boolean
    appState: string
    status: string
    app: {
        name: string
        version: string
    }
    stdout: string
    stderr: string
    stdoutLine: number
    stderrLine: number
    stdoutOldTop: number,
    stderrOldTop: number,
    reloadIntervalId: number
    promises: PromiseKeeper
}

export type StdElement = { scrollTop: number, scrollHeight: number } | null

export type MaxTime = {
    hours: number | null
    minutes: number | null
    seconds: number | null
} | null

export interface JobInfo {
    maxTime: MaxTime
    numberOfNodes: number | null
    tasksPerNode: number | null
}

export interface RunAppState {
    promises: PromiseKeeper
    error?: string
    loading: boolean
    appName: string
    favorite: boolean
    displayAppName: string
    appVersion: string
    appDescription: string
    appAuthor: string[]
    parameters: ApplicationParameter[]
    parameterValues: {}
    jobInfo: JobInfo
    tool: {} // ???
    comment: string
    jobSubmitted: boolean
}

export interface RunAppProps {
    match: match<{ appName: string, appVersion: string }>
    history: History
    updatePageTitle: () => void
}

export interface NumberParameter extends BaseParameter {
    defaultValue: { value: number, type: "double" | "int" } | null
    min: number | null
    max: number | null
    step: number | null
    type: "integer" | "floating_point"
}

export interface BooleanParameter extends BaseParameter {
    defaultValue: { value: boolean, type: "bool" } | null
    trueValue?: string | null
    falseValue?: string | null
    type: "boolean"
}

export interface InputFileParameter extends BaseParameter {
    defaultValue: string | null
    type: "input_file"
}

export interface InputDirectoryParameter extends BaseParameter {
    defaultValue: string | null
    type: "input_directory"
}

export interface TextParameter extends BaseParameter {
    defaultValue: { value: string, type: "string" } | null
    type: "text"
}

interface BaseParameter {
    name: string
    optional: boolean
    title: string
    description: string
    unitName?: string | null
}

export type ApplicationParameter = InputFileParameter | InputDirectoryParameter | NumberParameter | BooleanParameter | TextParameter

type Invocation = WordInvocation | VarInvocation

interface WordInvocation {
    type: "word"
    word: string
}

interface VarInvocation {
    type: "var"
    variableNames: string[]
    prefixGlobal: string
    suffixGlobal: string
    prefixVariable: string
    suffixVariable: string
    variableSeparator: string
}

type Info = { name: string, version: string }
export interface Description {
    info: Info
    tool: Info
    authors: string[]
    title: string
    description: string
    invocation: Invocation[]
    parameters: ApplicationParameter[]
    outputFileGlobs: [string, string]
}
interface Tool {
    owner: string
    createdAt: number
    modifiedAt: number
    description: ToolDescription
}

interface ToolDescription {
    info: Info
    container: string
    defaultNumberOfNodes: number,
    defaultTasksPerNode: number,
    defaultMaxTime: {
        hours: number
        minutes: number
        seconds: number
    }
    requiredModules: any[],
    authors: string[]
    title: string,
    description: string
    backend: string
}


export interface ApplicationInformation {
    owner: string
    favorite?: boolean
    createdAt, modifiedAt: number
    description: Description
    tool: Tool
}

export enum ParameterTypes {
    InputFile = "input_file",
    InputDirectory = "input_directory",
    Integer = "integer",
    FloatingPoint = "floating_point",
    Text = "text",
    Boolean = "boolean"
}

export interface SearchFieldProps {
    onSubmit: () => void
    icon: string
    placeholder: string
    value: string
    loading: boolean
    onValueChange: (value: string) => void
}