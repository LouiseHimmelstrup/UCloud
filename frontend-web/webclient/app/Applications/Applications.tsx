import * as React from "react";
import { Link } from "react-router-dom";
import * as Pagination from "Pagination";
import { Card, Icon, Rating, List } from "semantic-ui-react";
import { connect } from "react-redux";
import {
    fetchApplications,
    setLoading,
    updateApplications
} from "./Redux/ApplicationsActions";
import { updatePageTitle } from "Navigation/Redux/StatusActions";
import { Page } from "Types";
import { Application } from ".";
import { ApplicationsProps, ApplicationsOperations, ApplicationsStateProps } from ".";
import { setErrorMessage } from "./Redux/ApplicationsActions";
// Requires at least TS 3.0.0
import { MaterialColors } from "Assets/materialcolors.json";
import { favoriteApplicationFromPage } from "Utilities/ApplicationUtilities";
import { Cloud } from "Authentication/SDUCloudObject";
import { setPrioritizedSearch } from "Navigation/Redux/HeaderActions";

const COLORS_KEYS = Object.keys(MaterialColors);

// We need dynamic import due to nature of the import
const blurOverlay = require("Assets/Images/BlurOverlayByDan.png");

class Applications extends React.Component<ApplicationsProps> {

    componentDidMount() {
        const { props } = this;
        props.updatePageTitle();
        props.prioritizeApplicationSearch();
        props.setLoading(true);
        if (this.props.page.items.length === 0) {
            props.fetchApplications(props.page.pageNumber, props.page.itemsPerPage);
        }
    }

    render() {
        const { page, loading, fetchApplications, onErrorDismiss, updateApplications, error } = this.props;
        const favoriteApp = (app: Application) => updateApplications(favoriteApplicationFromPage(app, page, Cloud));
        return (
            <React.StrictMode>
                <Pagination.List
                    loading={loading}
                    onErrorDismiss={onErrorDismiss}
                    errorMessage={error}
                    onRefresh={() => fetchApplications(page.pageNumber, page.itemsPerPage)}
                    pageRenderer={({ items }: Page<Application>) =>
                        <Card.Group className="card-margin">
                            {items.map((app, index) => <SingleApplication key={index} app={app} favoriteApp={favoriteApp} />)}
                        </Card.Group>
                    }
                    page={page}
                    onItemsPerPageChanged={(size) => fetchApplications(0, size)}
                    onPageChanged={(pageNumber) => fetchApplications(pageNumber, page.itemsPerPage)}
                />
            </React.StrictMode >);
    }
}

interface SingleApplicationProps { app: Application, favoriteApp?: (app: Application) => void }
export function SingleApplication({ app, favoriteApp }: SingleApplicationProps) {
    const hashCode = toHashCode(app.description.info.name);
    const color = COLORS_KEYS[(hashCode % COLORS_KEYS.length)];
    const mClength = MaterialColors[color].length;
    const hex = MaterialColors[color][(hashCode % mClength)];
    const even = app.modifiedAt % 2 === 0;
    const opacity = even ? 0.3 : 1;
    const { description } = app.description;
    const image = even ? blurOverlay : `https://placekitten.com/200/200`;
    return (
        <Card>
            {/* FIXME: Move styling to .scss file where possible */}
            <div style={{
                background: hex
            }}>
                <Link to={`/appDetails/${app.description.info.name}/${app.description.info.version}/`}>
                    <div style={{
                        opacity: opacity,
                        width: "100%",
                        height: "200px",
                        backgroundImage: `url('${image}')`,
                        backgroundSize: "cover",
                        backgroundPosition: "center"
                    }} />
                </Link>
            </div>
            <Card.Content>
                <List horizontal floated="right">
                    {!!favoriteApp ? <List.Item>
                        <Rating icon={"star"} maxRating={1} rating={app.favorite ? 1 : 0} onClick={() => !!favoriteApp ? favoriteApp(app) : null} />
                    </List.Item> : null}
                    <List.Item>
                        <Link to={`/applications/${app.description.info.name}/${app.description.info.version}/`}>
                            <Icon color="green" name="play" />
                        </Link>
                    </List.Item>
                </List>
                <Card.Header
                    as={Link}
                    to={`/appDetails/${app.description.info.name}/${app.description.info.version}/`}
                    content={app.description.title}
                />
                <Card.Meta content={app.description.info.version} />
            </Card.Content>
            <Card.Content extra>
                {description.length > 72 ? `${description.slice(0, 72)}...` : description}
            </Card.Content>
        </Card>
    );
}

function toHashCode(name: string): number {
    let hash = 0;
    if (name.length == 0) {
        return hash;
    }
    for (let i = 0; i < name.length; i++) {
        let char = name.charCodeAt(i);
        hash = ((hash << 5) - hash) + char;
        hash = hash & hash; // Convert to 32bit integer
    }
    return Math.abs(hash);
}

const mapDispatchToProps = (dispatch): ApplicationsOperations => ({
    prioritizeApplicationSearch: () => dispatch(setPrioritizedSearch("applications")),
    onErrorDismiss: () => dispatch(setErrorMessage()),
    updatePageTitle: () => dispatch(updatePageTitle("Applications")),
    setLoading: (loading: boolean) => dispatch(setLoading(loading)),
    fetchApplications: (pageNumber: number, itemsPerPage: number) => dispatch(fetchApplications(pageNumber, itemsPerPage)),
    updateApplications: (applications: Page<Application>) => dispatch(updateApplications(applications))
});

const mapStateToProps = ({ applications }): ApplicationsStateProps => ({
    favCount: applications.page.items.filter(it => it.favorite).length,
    ...applications
});

export default connect(mapStateToProps, mapDispatchToProps)(Applications);