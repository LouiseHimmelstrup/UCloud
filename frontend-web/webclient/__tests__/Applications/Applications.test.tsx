import * as React from "react";
import * as Renderer from "react-test-renderer";
import { ApplicationCard } from "Applications/Card";
import { configureStore } from "Utilities/ReduxUtilities";
import { init } from "Applications/Redux/BrowseObject";
import applicationsReducer from "Applications/Redux/BrowseReducer";
import { applicationsPage } from "../mock/Applications";
import { MemoryRouter } from "react-router";
import { shallow } from "enzyme";

const emptyPageStore = configureStore({ applications: init() }, { applications: applicationsReducer });
const fullPageStore = {
    ...emptyPageStore
};



describe("Single Application Component", () => {
    test.skip("Render Single Application", () => {
        expect(Renderer.create(
            <MemoryRouter>
                <ApplicationCard
                    app={applicationsPage.items[0]}
                    onFavorite={() => null}
                />
            </MemoryRouter>).toJSON()).toMatchSnapshot();
    });
});

describe("Single Applications", () => {

    test.skip("Favorite application", () => {
        const func = jest.fn();
        const application = applicationsPage.items[0];
        const singleApp = shallow(<ApplicationCard
            app={application}
            onFavorite={func}
        />);
        singleApp.simulate("click");
        expect(func).toHaveBeenCalled();
    });
});