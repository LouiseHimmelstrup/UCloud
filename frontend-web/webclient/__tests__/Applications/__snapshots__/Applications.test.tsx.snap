// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Single Application Component Render Single Application 1`] = `
<div
  className="sc-kGXeez cABhMr"
  height={212}
  width={252}
>
  <div
    className="sc-gipzik hYhsek"
  >
    <div
      className="sc-bdVaJa bQBfBb"
    >
      <div
        className="sc-bdVaJa bQBfBb"
        style={
          Object {
            "background": "#E91E63",
          }
        }
      >
        <img
          className="sc-hMqMXs iipvbU"
          src={[Function]}
          style={
            Object {
              "opacity": 0.4,
            }
          }
        />
      </div>
      <div
        className="sc-bwzfXH hDECRi"
      >
        <div
          className="sc-EHOje jjZauP"
          color="white"
          cursor="inherit"
          fontSize={2}
        >
          PALMS
        </div>
      </div>
      <div
        className="sc-bwzfXH iLsITc"
      >
        <div
          className="sc-EHOje irgxuc"
          color="white"
          cursor="inherit"
          fontSize="xxs-small"
        >
          v 
          1.0.0
        </div>
      </div>
      <div
        className="sc-bwzfXH hTtHjt"
      >
        <svg
          className="sc-gZMcBi gyTeqP sc-iwsKbI jzxgXF"
          color="red"
          cursor="pointer"
          fill="currentcolor"
          height={24}
          onClick={[Function]}
          viewBox="0 0 23 22"
          width={24}
        >
          <path
            d="M14.939 7.146l7.886 1.146-5.707 5.562 1.347 7.854L11.412 18 4.36 21.708l1.347-7.854L0 8.292l7.886-1.146L11.412 0l3.527 7.146zM9.878 9.888l-3.432.498 2.483 2.42-.586 3.418 3.07-1.614 3.068 1.614-.586-3.417 2.483-2.42-3.431-.499-1.535-3.11-1.534 3.11z"
            fill="#c00"
          />
        </svg>
      </div>
      <div
        className="sc-bwzfXH biWzgY"
      >
        <div
          className="sc-dnqmqq gAvPes"
          color="white"
          title="by Dan Sebastian Thrane <dthrane@imada.sdu.dk>"
          width={180}
        >
          by 
          Dan Sebastian Thrane &lt;dthrane@imada.sdu.dk&gt;
        </div>
      </div>
      <div
        className="sc-bwzfXH cebrxn"
      >
        <a
          className="sc-cMljjf gvnKqJ"
          color="black"
          href="/applications/palms/1.0.0/"
          onClick={[Function]}
          theme={
            Object {
              "bold": 400,
              "boxShadows": Array [
                "0 0 2px 0 rgba(0,0,0,.08),0 1px 4px 0 rgba(0,0,0,.16)",
                "0 0 2px 0 rgba(0,0,0,.08),0 2px 8px 0 rgba(0,0,0,.16)",
                "0 0 2px 0 rgba(0,0,0,.08),0 4px 16px 0 rgba(0,0,0,.16)",
                "0 0 2px 0 rgba(0,0,0,.08),0 8px 32px 0 rgba(0,0,0,.16)",
              ],
              "breakpoints": Array [
                "32em",
                "40em",
                "48em",
                "64em",
              ],
              "colorStyles": Object {
                "blueOnLightBlue": Object {
                  "backgroundColor": "#cdf",
                  "color": "#0055d5",
                },
                "blueOnWhite": Object {
                  "backgroundColor": "#fff",
                  "color": "#0055d5",
                },
                "danger": Object {
                  "backgroundColor": "#c00",
                  "color": "#fff",
                },
                "grayOnWhite": Object {
                  "backgroundColor": "#fff",
                  "color": "#c9d3df",
                },
                "greenOnLightGreen": Object {
                  "backgroundColor": "#cec",
                  "color": "#0a0",
                },
                "greenOnWhite": Object {
                  "backgroundColor": "#fff",
                  "color": "#0a0",
                },
                "info": Object {
                  "backgroundColor": "#ebeff3",
                  "color": "#001833",
                },
                "purpleOnLightPurple": Object {
                  "backgroundColor": "#ecf",
                  "color": "#70b",
                },
                "purpleOnWhite": Object {
                  "backgroundColor": "#fff",
                  "color": "#70b",
                },
                "redOnLightRed": Object {
                  "backgroundColor": "#fcc",
                  "color": "#c00",
                },
                "redOnWhite": Object {
                  "backgroundColor": "#fff",
                  "color": "#c00",
                },
                "success": Object {
                  "backgroundColor": "#0a0",
                  "color": "#fff",
                },
                "textOnLightGray": Object {
                  "backgroundColor": "#ebeff3",
                  "color": "#001833",
                },
                "textOnOrange": Object {
                  "backgroundColor": "#fa0",
                  "color": "#001833",
                },
                "textOnWhite": Object {
                  "backgroundColor": "#fff",
                  "color": "#001833",
                },
                "warning": Object {
                  "backgroundColor": "#fa0",
                  "color": "#001833",
                },
                "whiteOnBlue": Object {
                  "backgroundColor": "#0055d5",
                  "color": "#fff",
                },
                "whiteOnDarkOrange": Object {
                  "backgroundColor": "#a50",
                  "color": "#fff",
                },
                "whiteOnGray": Object {
                  "backgroundColor": "#c9d3df",
                  "color": "#fff",
                },
                "whiteOnGreen": Object {
                  "backgroundColor": "#0a0",
                  "color": "#fff",
                },
                "whiteOnPurple": Object {
                  "backgroundColor": "#70b",
                  "color": "#fff",
                },
                "whiteOnRed": Object {
                  "backgroundColor": "#c00",
                  "color": "#fff",
                },
                "whiteOnText": Object {
                  "backgroundColor": "#001833",
                  "color": "#fff",
                },
              },
              "colors": Object {
                "black": "#000",
                "blue": "#0055d5",
                "borderGray": "#d1d6db",
                "darkBlue": "#049",
                "darkGray": "#53657d",
                "darkGreen": "#060",
                "darkOrange": "#a50",
                "darkRed": "#800",
                "gray": "#c9d3df",
                "green": "#0a0",
                "lightBlue": "#cdf",
                "lightGray": "#ebeff3",
                "lightGreen": "#cec",
                "lightPurple": "#ecf",
                "lightRed": "#fcc",
                "midGray": "#53657d",
                "orange": "#fa0",
                "paginationHoverColor": "#f7f7f7",
                "purple": "#70b",
                "red": "#c00",
                "text": "#001833",
                "white": "#fff",
              },
              "duration": Object {
                "fast": "150ms",
                "normal": "300ms",
                "slow": "450ms",
                "slowest": "600ms",
              },
              "fontFamily": "'IBM Plex Sans',sans-serif",
              "fontSizes": Array [
                12,
                14,
                16,
                20,
                24,
                32,
                40,
                56,
                72,
              ],
              "fontWeights": Object {
                "bold": 400,
                "medium": 300,
                "regular": 300,
              },
              "letterSpacings": Object {
                "caps": "0.025em",
                "normal": "normal",
              },
              "lineHeights": Object {
                "display": 1.25,
                "standard": 1.5,
              },
              "maxContainerWidth": "1280px",
              "mediaQueries": Array [
                "@media screen and (min-width:32em)",
                "@media screen and (min-width:40em)",
                "@media screen and (min-width:48em)",
                "@media screen and (min-width:64em)",
              ],
              "radii": Array [
                0,
                2,
                6,
              ],
              "radius": "2px",
              "regular": 300,
              "space": Array [
                0,
                4,
                8,
                16,
                32,
                64,
                128,
              ],
              "textStyles": Object {
                "body0": Object {
                  "fontSize": "12px",
                  "fontWeight": 300,
                  "lineHeight": 1.5,
                },
                "body1": Object {
                  "fontSize": "14px",
                  "fontWeight": 300,
                  "lineHeight": 1.5,
                },
                "body2": Object {
                  "fontSize": "16px",
                  "fontWeight": 300,
                  "lineHeight": 1.5,
                },
                "display0": Object {
                  "fontSize": "12px",
                  "fontWeight": 400,
                  "letterSpacing": "0.025em",
                  "lineHeight": 1.25,
                  "textTransform": "uppercase",
                },
                "display1": Object {
                  "fontSize": "14px",
                  "fontWeight": 400,
                  "lineHeight": 1.25,
                },
                "display2": Object {
                  "fontSize": "16px",
                  "fontWeight": 400,
                  "lineHeight": 1.25,
                },
                "display3": Object {
                  "fontSize": "20px",
                  "fontWeight": 400,
                  "lineHeight": 1.25,
                },
                "display4": Object {
                  "fontSize": "24px",
                  "fontWeight": 400,
                  "lineHeight": 1.25,
                },
                "display5": Object {
                  "fontSize": "32px",
                  "fontWeight": 400,
                  "lineHeight": 1.25,
                },
                "display6": Object {
                  "fontSize": "40px",
                  "fontWeight": 400,
                  "lineHeight": 1.25,
                },
                "display7": Object {
                  "fontSize": "56px",
                  "fontWeight": 400,
                  "lineHeight": 1.25,
                },
                "display8": Object {
                  "fontSize": "72px",
                  "fontWeight": 400,
                  "lineHeight": 1.25,
                },
              },
              "timingFunctions": Object {
                "easeIn": "cubic-bezier(0.5, 0, 1, 1)",
                "easeInOut": "cubic-bezier(0.5, 0, 0.25, 1)",
                "easeOut": "cubic-bezier(0, 0, 0.25, 1)",
              },
              "transitionDelays": Object {
                "large": "260ms",
                "medium": "160ms",
                "small": "60ms",
                "xLarge": "360ms",
              },
            }
          }
        >
          <svg
            className="sc-dxgOiQ jQZQST sc-iwsKbI fSJFTj"
            cursor="pointer"
            fill="currentcolor"
            height={38}
            viewBox="0 0 32 36"
            width={38}
          >
            <path
              d="M0 2.04C0 1.3.37.65 1.02.28c.65-.37 1.4-.37 2.04.02l27.27 15.88c.64.37 1 1 1 1.74 0 .73-.36 1.38-1 1.75L3.06 35.55a2 2 0 0 1-2.04.01A1.98 1.98 0 0 1 0 33.8V2.04z"
              fill="#8393a7"
            />
          </svg>
        </a>
      </div>
    </div>
  </div>
  <div
    className="sc-gipzik hYhsek"
  >
    <div
      className="sc-bwzfXH bVqAki"
    >
      <div
        className="sc-EHOje kvayUm"
        cursor="inherit"
      >
        The central role of places in which physical activity (PA) is done is now widely recognized, so it i
      </div>
    </div>
  </div>
</div>
`;
